docs/architecture/adr/ADR045-dev-namespaces.md

```diff
@@ -0,0 +1,83 @@
+# ADR045: Dev namespaces
+
+## Status
+
+Pending
+
+## Context
+
+Following the retirement of gsp-local the service teams have no way of testing
+changes before a merge to master. So the feedback cycle is very slow, and
+potentially risky. We need to enable devs to get more rapid feedback through use
+of namespaces that allow for deployments from arbitrary branches or from local
+`kubectl apply` actions.
+
+This presents several security problems that need to be addressed.
+
+The in-cluster concourse:
+
+* must not be able to create or edit daemonsets
```

Comment:
What's the security risk with DaemonSets?

Reply:
They are a "one per node" deployment model, which we don't want tenants to be able to manage / manipulate.

Reply:
To add my tuppence: I don't think there's much security risk with daemonsets per se, but I also don't think they're a thing we want to expose to tenants.

In particular: a tenant could, if they wanted, deliberately run pods with affinity rules to run on each node.  They could then wrap that up in a custom controller.  In effect, even if we ban daemonsets, users have all the tools to make their own.

On the other hand, most of the valid use cases of DaemonSets imply some sort of extra privileges.  For example, our CNIs run as DaemonSets, as do kiam-agent and kiam-server.  I can't think of something you'd actually need a DaemonSet for that didn't have any extra privileges of some sort.

So while I don't think it's actually required that we ban DaemonSets, I think it's better if tenants can't use them and I can't see any need for tenants to use them.

Reply:
Okay, let's word this to that effect?

Reply:
I've added a few notes, but I don't really want to go into too much detail in a bulleted list.